cl/users/models.py

```diff
@@ -190,8 +259,211 @@ def generate_recap_email(user_profile: UserProfile, append: int = None) -> str:
+) -> None:
+    """Handle a soft bounce notification received from SNS
+    :param message_id: The unique message id assigned by Amazon SES
+    :param bounceSubType: The subtype of the bounce, as determined
+     by Amazon SES
+    :param recipient_emails: a list of email addresses one per recipient
+     to whom the bounce notification pertains
+    :return: None
+    """
+
+    back_off_events = ["Undetermined", "General", "MailboxFull"]
+    small_only_events = ["MessageTooLarge", "AttachmentRejected"]
+
+    MAX_RETRY_COUNTER = 5
+    INITIAL_HOURS = 2
+
+    for email in recipient_emails:
+        if bounceSubType in back_off_events:
+            # Handle events that must trigger a backoff event
+            backoff_event = BackoffEvent.objects.filter(
+                email_address=email,
+            )
+            if not backoff_event.exists():
+                # Create a backoff event for an email address if not exists
+                # Initialize retry_counter and next_retry_date
+                next_retry_date = now() + timedelta(hours=INITIAL_HOURS)
+                BackoffEvent.objects.create(
+                    email_address=email,
+                    retry_counter=0,
+                    next_retry_date=next_retry_date,
+                )
+            else:
+                # If a previous backoff event exists
+                retry_counter = backoff_event[0].retry_counter
+                next_retry_date = backoff_event[0].next_retry_date
+
+                # Check if waiting period expired
+                if next_retry_date <= now():
+                    if retry_counter >= MAX_RETRY_COUNTER:
+                        # Check if backoff event has reached
+                        # max number of retries, if so ban email address
+                        # Only ban email address if not previously banned
+                        EmailFlag.objects.get_or_create(
+                            email_address=email,
+                            flag_type="ban",
+                            defaults={
+                                "flag": "max_retry_reached",
+                                "reason": bounceSubType,
+                            },
+                        )
+                    else:
+                        # If max number of retries has not been reached,
+                        # update backoff event, update retry_counter
+                        new_retry_counter = retry_counter + 1
+                        # Update new_next_retry_date exponentially
+                        new_next_retry_date = now() + timedelta(
+                            hours=pow(INITIAL_HOURS, new_retry_counter + 1)
+                        )
+                        BackoffEvent.objects.filter(
+                            email_address=email,
+                        ).update(
+                            retry_counter=new_retry_counter,
+                            next_retry_date=new_next_retry_date,
+                        )
```

Comment:
There's a race condition here, but I suppose it's fine. Easier to live with it than to fix it, probably.

Reply:
I see, I definitely would like to fix this, would it be helpful to use `select_for_update()` and `@transaction.atomic`?
I would be very grateful if you could give me a hint as to whether this is a good approach or if there might be a better one.

Reply:
Yes, I think `select_for_update` would be a simple and good way to do this.